File: javascript/src/repository/user.cockroach.js

export default class UserCockroachdbRepository {
  constructor(pool) {
    this._pool = pool;
  }

  async getById(id) {
    return await withTransaction(this._pool, async client => {
      const result = await client.query(
        "SELECT * FROM users WHERE id = $1",
        [id]);

      return this._toUser(result.rows[0]);
    });
  }

  async create(user) {
    return await withTransaction(this._pool, async client => {
      const result = await client.query(
        "INSERT INTO users (username, name, age, surname, password, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *",
        [user.username, user.name, user.age, user.surname, user.password, new Date(), new Date()]);

      return this._toUser(result.rows[0])
    });
  }

  _toUser(row) {
    const user = {
      ...row,
      createdAt: row.created_at,
      updatedAt: row.updated_at
    }
    delete user.created_at;
    delete user.updated_at;
    return user;
  }

}

// Making it simpler
function withTransaction(pool, operation) {
  return new Promise(async (resolve, reject) => {
    const client = await pool.connect();
    retryTxn(0, 15, client,
      async (c, cb) => {
        const result = await operation(c);
        cb(null, result);
      }, async (err, res) => {
        await client.release();
        if (err) {
          reject(err);
        } else {
          resolve(res);
        }
      })
  });
}

// Based on https://www.cockroachlabs.com/docs/stable/build-a-nodejs-app-with-cockroachdb.html?filters=local#step-2-get-the-code
async function retryTxn(n, max, client, operation, callback) {
  await client.query("BEGIN;");
  while (true) {
    n++;
    if (n === max) {
      throw new Error("Max retry count reached.");
    }
    try {
      await operation(client, callback);
      await client.query("COMMIT;");
      return;
    } catch (err) {
      if (err.code !== "40001") {
        return callback(err);
      } else {
        console.log("Transaction failed. Retrying transaction.");
        console.log(err.message);
        await client.query("ROLLBACK;", () => {
          console.log("Rolling back transaction.");
        });
        await new Promise((r) => setTimeout(r, 2 ** n * 1000));
      }
    }
  }
}

File: golang/repository/user.cockroach.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/bigmontz/nothing/repository/almost_crdb"
	"strconv"
	"time"
)

type userCockRoachRepository struct {
	db *sql.DB
}

func NewUserCockroachRepository(db *sql.DB) UserRepository {
	return &userCockRoachRepository{db: db}
}

func (u *userCockRoachRepository) Create(user *User) (*User, error) {
	now := time.Now().In(time.UTC)
	row := u.db.QueryRow(
		"INSERT INTO users (username, name, age, surname, password, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *",
		user.Username,
		user.Name,
		user.Age,
		user.Surname,
		user.Password,
		now,
		now)

	if err := row.Err(); err != nil {
		return nil, err
	}
	var result User
	if err := row.Scan(
		&result.Id,
		&result.Username,
		&result.Name,
		&result.Surname,
		&result.Password,
		&result.Age,
		&result.CreatedAt,
		&result.UpdatedAt); err != nil {
		return nil, err
	}
	return &result, nil
}

func (u *userCockRoachRepository) FindById(rawUserId interface{}) (*User, error) {
	userId, err := strconv.Atoi(rawUserId.(string))
	if err != nil {
		return nil, userError{err: fmt.Errorf("invalid user ID: %w", err)}
	}
	row := u.db.QueryRow("SELECT * FROM users WHERE id = $1", userId)
	if err := row.Err(); err != nil {
		return nil, err
	}
	var result User
	if err := row.Scan(
		&result.Id,
		&result.Username,
		&result.Name,
		&result.Surname,
		&result.Password,
		&result.Age,
		&result.CreatedAt,
		&result.UpdatedAt); err != nil {
		return nil, err
	}
	return &result, nil
}

func (u *userCockRoachRepository) UpdatePassword(userId interface{}, passwordUpdate *PasswordUpdate) (*User, error) {
	ctx := context.Background()
	result, err := almost_crdb.SlightlyAdaptedExecuteTx(
		ctx,
		u.db,
		nil,
		func(tx almost_crdb.SlightlyAdaptedTx) (interface{}, error) {
			result, err := tx.SlightlyAdaptedExec(
				ctx,
				"UPDATE users SET password = $3, updated_at = $4 WHERE id = $1 AND password = $2",
				userId,
				passwordUpdate.Current,
				passwordUpdate.New,
				time.Now().In(time.UTC),
			)
			if err != nil {
				return nil, err
			}
			rowsAffected, err := result.RowsAffected()
			if err != nil {
				return nil, err
			}
			if rowsAffected == 0 {
				return nil, userError{err: fmt.Errorf("user not found"), notFound: true}
			}
			return &User{Id: userId}, nil
		},
	)
	if err != nil {
		return nil, err
	}
	return result.(*User), nil
}

func (u *userCockRoachRepository) Close() error {
	return u.db.Close()
}
